server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java

```diff
@@ -1290,36 +1281,92 @@ private void ensureMutable() {
      * all the nodes have been returned.
      */
-    public Iterator<ShardRouting> nodeInterleavedShardIterator() {
+    public Iterator<ShardRouting> nodeInterleavedShardIterator(boolean movePrimaryFirst) {
         final Queue<Iterator<ShardRouting>> queue = new ArrayDeque<>();
         for (Map.Entry<String, RoutingNode> entry : nodesToShards.entrySet()) {
             queue.add(entry.getValue().copyShards().iterator());
         }
-        return new Iterator<ShardRouting>() {
-            public boolean hasNext() {
-                while (!queue.isEmpty()) {
-                    if (queue.peek().hasNext()) {
+        if (movePrimaryFirst) {
+            return new Iterator<ShardRouting>() {
+                private Queue<ShardRouting> replicaShards = new ArrayDeque<>();
+                private Queue<Iterator<ShardRouting>> replicaIterators = new ArrayDeque<>();
+
+                public boolean hasNext() {
+                    while (!queue.isEmpty()) {
+                        if (queue.peek().hasNext()) {
+                            return true;
+                        }
+                        queue.poll();
+                    }
+                    if (!replicaShards.isEmpty()) {
                         return true;
                     }
-                    queue.poll();
+                    while (!replicaIterators.isEmpty()) {
+                        if (replicaIterators.peek().hasNext()) {
+                            return true;
+                        }
+                        replicaIterators.poll();
+                    }
+                    return false;
                 }
-                return false;
-            }
 
-            public ShardRouting next() {
-                if (hasNext() == false) {
-                    throw new NoSuchElementException();
+                public ShardRouting next() {
+                    if (hasNext() == false) {
+                        throw new NoSuchElementException();
+                    }
+                    while (!queue.isEmpty()) {
+                        Iterator<ShardRouting> iter = queue.poll();
+                        if (iter.hasNext()) {
+                            ShardRouting result = iter.next();
+                            if (result.primary()) {
+                                queue.offer(iter);
+                                return result;
+                            }
+                            replicaShards.offer(result);
+                            replicaIterators.offer(iter);
+                        }
+                    }
+                    if (!replicaShards.isEmpty()) {
+                        return replicaShards.poll();
+                    }
+                    Iterator<ShardRouting> replicaIterator = replicaIterators.poll();
+                    ShardRouting replicaShard = replicaIterator.next();
+                    replicaIterators.offer(replicaIterator);
+                    return replicaShard;
                 }
```

Comment:
For my understanding, is it possible for replicaIterators to contain primary shards, since on the first replica that we encounter for a node we are transferring the iterator over to `replicaIterator `?

Reply:
It is not possible for replica iterators to contain primary shards as RoutingNode iterates over primary shards first